woocommerce/woocommerce-android
# Convert fetchProductReviews into suspendable function

Merge Instructions:
1. Review [the PR in FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2188)
2. Review this PR
3. Merge FluxC
4. Update FluxC hash
5. Remove "Do not merge" label
5. Merge this PR

This PR refactors fetchProductReviews into a suspendable function - removes usage of the event bus.

There are two reasons for this change

1. [The event bus architecture is considered as legacy/deprecated](https://github.com/wordpress-mobile/WordPress-FluxC-Android/wiki/%5BDeprecated%5D-Architecture)
2. WCAndroid app uses stateful repositories which need to be registered to the event bus - this leads to memory leaks when the repository doesn't unregister from the event bus. It also causes side-effects when multiple instances of a repository are created - eg. duplicate tracking events.

Note: I recommend reviewing this PR by commits to better follow my train of thoughts.

Other changes - 
1. Enhancement `ProductReviewsRepository.kt `: Returns result instead of cache data from `fetchApprovedProductReviewsFromApi` -> the method is prefixed with `fetch` and returning cached data can be misleading.
2. Enhancement `ProductReviewsRepository.kt `: Moves tracking to VM
3. Fix `ProductReviewsRepository.kt `: Passes `loadMore` to tracking instead of `isLoadingMore` which was always false
4. Fix `ProductReviewsRepository.kt `: Updates page offset only when the request succeeds - previously the offset was updated even when the request failed. (the first commit introduced an issue fixed in https://github.com/woocommerce/woocommerce-android/pull/5207/commits/e5d99e179b82d501eff21fced8ac2abf9645c561)
5. Enhancement `ProductReviewsRepository.kt `: Shows snackbar with an error on failure instead of silently reloading data from cache
6. Fix `ReviewListRepository.kt`: Updates page offset only when the request succeeds - previously the offset was updated even when the request failed. (the first commit introduced an issue fixed in https://github.com/woocommerce/woocommerce-android/pull/5207/commits/e5d99e179b82d501eff21fced8ac2abf9645c561)


To Test:
1. Clean app data and login
2. Open Reviews List
2. Make sure the reviews get loaded
3. Scroll down and make sure more reviews get loaded
4. Verify `REVIEWS_LOADED` events gets tracked (in LogCat - if you don't see it make sure to enable logging in privacy settings of the app)
5. Go to a detail of a product with reviews
6. Make sure the reviews get loaded
7. Publish a new review in a browser
8. Pull to refresh in the app and make sure the review appears (if it doesn't appear approve it on the web)

~P.S. I tested this with a site with just a few reviews. I asked if someone could invite me to a site with more reviews, if no-one invites me, I'll somehow generate them and re-test it.~ (tested with 100+ reviews on the site)
